### ISSUE TITLE
emacs issues

### ISSUE THREAD

**Issue by evangreen** [minoca/os#74]
As reported by Richard in his email:
(display-time) causes a segfault.

dired fails because df is missing (the statvfs issue).

Key binding is screwed... Ctrl-space should be bound to set-mark-command,
but is not recognized.  Terminal interface and control may be an area
that needs much work in Minoca.

shell is workable, but has issues in this area - complains "Failed to set
raw input mode." after every input line, unwanted echo.

**Comment by drawkula** [comment]
`emacs`' display sometimes is screwed and correct after typing `ctrl-L`.

It is easy to see with indented source code, which sometimes seems to have wrongly indented lines being in the right place after `ctrl-L`. This may be a terminal emulation related instead of a `emacs` problem or bug.

Example:

Some mis-indented lines...

![20161203-135204-gmt--minoca-os--emacs-screen-glitches-1a](https://cloud.githubusercontent.com/assets/9768622/20859758/6583d1ea-b968-11e6-8911-1828f619c208.png)

After hitting `ctrl-L` (twice):

![20161203-135319-gmt--minoca-os--emacs-screen-glitches-1b](https://cloud.githubusercontent.com/assets/9768622/20859765/942a5c58-b968-11e6-917c-683afdc21711.png)

**Comment by ryniker** [comment]
Just a hack to get dired working...

Great improvement for a trivial effort.  I never thought to fudge.  Thank you.

**Comment by drawkula** [comment]
It is hacking in/as/for self defense!

That's ok... :-)

**Comment by ryniker** [comment]
display-time works fine in the emacs version 25-1 that I built on my Raspberry Pi.
I suspect the problem lies not in the emacs code, but in the build operation that created the package in third-party.

I wrote some notes about what I did to build emacs-25.1 in
http://ryniker.org/minoca/emacs-25.1_build_notes

**Comment by ryniker** [comment]
The shell problems come from swiss (/bin/sh).  The function SwSetRawInputMode in
os/apps/swiss/swlib/uos.c emits the message "Failed to set raw input
mode."  The reason for this is unknown, but it probably also accounts for
the undesired echo of input lines.  Solution for both issues: install
the bash package, and customize the emacs variable shell-file-name to
"/usr/bin/bash" for a much more pleasant experience.

**Comment by ryniker** [comment]
Even with Evan's suggestion in Issue #60 to copy /bin/swiss and make it setuid root, su does not
work in an emacs shell buffer:

bash-4.3$ su
su: Incorrect password.
bash-4.3$ 

This presumably involves a no-echo read of a password in the mini-buffer, and needs further investigation.

**Comment by evangreen** [comment]
Okay, I've made some changes to both the os repository and emacs to fix these issues. To try this out, there are two options:
* Option one: grab my privately compiled emacs from [here](https://www.minocacorp.com/packages/0.3/i686/test/emacs_24.5_minoca-i686.ipk), and rebuild the os from source.
* Option two: Pull the latest os, re-run `make tools` in third-party (because there are C library header changes), then rebuild emacs from source in third-party (with a `make clean` in third-party/build/emacs first).

You'll need the latest from the os repository. Things that should be fixed:
* display-time segfault ac6e9aa1c3a130f57d892c05d8d34738b955402d
* gracefully handle missing dired [a8d24f97](https://gitlab.com/minoca/third-party/commit/a8d24f971af62a48cd38a04550f833afe1856812)
* Control-space issues 7d9fbcd47c888fb4599d8143a25a69d26784aaee
* Fix for "Failed to set raw input mode" 4eaab353b8b6a502ba45643357766a33b8979f7f
* Fix for emacs using pipe instead of pty on non-root accounts 4eaab353b8b6a502ba45643357766a33b8979f7f
* Fix for emacs thinking shell exited with signal 127 27c5e92afca30545403cc62c7b26341781fcd99e
* Fix for bad indentation 60db3f87c06e2aa5321849bc0f1f67f116cad85b

Basically, the emacs shell stuff should work now with the exception that there's an extra echo of the command.

**Comment by ryniker** [comment]
Using GNU Emacs 24.5.1 (arm-none-minoca) of 2016-11-06 on minoca (installed by opkg) an attempt to create a shell buffer still produces:
Process shell unknown signal 127 (core dumped)
This does not occur when I use the later version I built: GNU Emacs 25.1.1 (armv6-unknown-linux-gnu) of 2016-12-20
Both trials on: Minoca minoca 0.3.1.2039-development-debug ryniker-05ab527 Dec 16 2016 20:25:38 armv6

**Comment by evangreen** [comment]
Thanks Richard. I haven't pushed the new binaries to the server yet, but will do that soon once we get some clean builds.

**Comment by evangreen** [comment]
I believe these issues should be fixed up in the latest 0.4 binaries.
